frontend/simple/js/state.js

```diff
@@ -5,22 +5,123 @@
 
 // TODO: save only relevant state to localforage
 const state = {
+  currentGroupLog: null,
+  availableGroups: [],
   // TODO: this should be managed by Keychain, not here
-  loggedIn: false
+  loggedInUser: null,
+  offset: [],
+  socket: null
 }
 
 // Mutations must be synchronous! Never call these directly!
 // http://vuex.vuejs.org/en/mutations.html
 const mutations = {
-  LOGIN (state) { state.loggedIn = true },
-  LOGOUT (state) { state.loggedIn = false }
+  login (state, user) { state.loggedInUser = user },
+  logout (state) { state.loggedInUser = null },
+  updateCurrentGroupLog (state, log) {
+    if (Array.isArray(log.offset)) {
+      state.currentGroupLog = log.log
+      state.offset = log.offset
+    } else {
+      state.currentGroupLog = log
+    }
+  },
+  updateAvailableGroups (state, available) {
+    state.availableGroups = available
+  },
+  updateSocket (state, socket) {
+    state.socket = socket
+  }
+}
+
+export const actions = {
+  async createGroup ({commit, state}, group) {
+    let log = await EventLog.createEventLogFromGroup(group)
+    let available = state.availableGroups
+    if (state.socket) {
+      await joinRoom(state.socket, group)
+    }
+    commit('updateCurrentGroupLog', log)
+    available.push(log.currentLogPosition)
+    commit('updateAvailableGroups', available)
+  },
+  // TODO: Connect this to user credentials
+  // limits to available groups for user
+  async loadGroup ({commit, state}, group) {
+    let available = state.availableGroups
+    if (available.find((grp) => { return group === grp })) {
+      if (state.socket) {
+        await joinRoom(state.socket, group)
+      }
+      let log = await EventLog.getEventLogForGroupId(group)
+      let offset = []
+      commit('updateCurrentGroupLog', {log, offset})
+    }
+  },
+  async appendLog ({commit, state}, event) {
+    let log = await EventLog.addItemToLog(state.currentGroupLog, event)
+    commit('updateCurrentGroupLog', log)
+  },
+  async receiveEvent ({commit, state}, msg) {
+    // check if this is a log update
+    if (msg.data && msg.data.entry && msg.data.entry.parentHash) {
+      let available = state.availableGroups
+      // check to see if this event is for an available group
+      if (available.find((grp) => { return msg.data.groupId === grp })) {
+        let log = await EventLog.getEventLogForGroupId(msg.data.groupId)
+        log = await EventLog.addItemToLog(log, msg.data.entry.data, true)
+        // update curren.t if the event is for the current group
+        if (state.currentGroupLog && state.currentGroupLog.currentLogPosition === msg.data.groupId) {
+          commit('updateCurrentGroupLog', log)
+        }
+      }
+    }
+  },
+  async backward ({commit, state}) {
+    if (state.currentGroupLog) {
+      let entry = await EventLog.getItemFromLog(state.currentGroupLog, state.currentGroupLog.currentLogPosition)
+      let log = {groupId: state.currentGroupLog.groupId, currentLogPosition: entry.parentHash}
```

Comment:
Since a log is a type (and not just random data), use a function to create it instead of an inline object. The idea is that it's "like an object instance", and like classes and objects, constructors are used to create them. This way if the structure of the object ever changes, you don't have to modify it in 20 different places throughout the project.